oak_attestation_verification/src/amd.rs

```diff
@@ -0,0 +1,202 @@
+//
+
+//! Contains structs for specifying in-toto statements and claims about
+
+use alloc::string::String;
+
+use oak_sev_guest::guest::{AttestationReport, TcbVersion};
+use p256::pkcs8::ObjectIdentifier;
+use rsa::{pss::Signature, signature::Verifier, RsaPublicKey};
+use sha2::Sha384;
+use x509_cert::{
+    der::{referenced::OwnedToRef, Encode},
+    Certificate, Version,
+};
+use zerocopy::{AsBytes, FromZeroes};
+
+const RSA_SSA_PSS_OID: ObjectIdentifier = ObjectIdentifier::new_unwrap("1.2.840.113549.1.1.10");
+const PRODUCT_NAME_OID: ObjectIdentifier = ObjectIdentifier::new_unwrap("1.3.6.1.4.1.3704.1.2");
+const BL_OID: ObjectIdentifier = ObjectIdentifier::new_unwrap("1.3.6.1.4.1.3704.1.3.1");
+const TEE_OID: ObjectIdentifier = ObjectIdentifier::new_unwrap("1.3.6.1.4.1.3704.1.3.2");
+const SNP_OID: ObjectIdentifier = ObjectIdentifier::new_unwrap("1.3.6.1.4.1.3704.1.3.3");
+const CHIP_ID_OID: ObjectIdentifier = ObjectIdentifier::new_unwrap("1.3.6.1.4.1.3704.1.4");
+const UCODE_OID: ObjectIdentifier = ObjectIdentifier::new_unwrap("1.3.6.1.4.1.3704.1.3.8");
+
+// Verifies validity of a matching ARK, ASK certificate pair.
+//
+// Validate at least a subset of Appendix B.3 of
+// https://www.amd.com/content/dam/amd/en/documents/epyc-technical-docs/programmer-references/55766_SEV-KM_API_Specification.pdf
+// Ideally, we'd check everything listed there.
+pub fn validate_ark_ask_certs(ark: &Certificate, ask: &Certificate) -> anyhow::Result<()> {
+    if ark.tbs_certificate.version != Version::V3 {
+        anyhow::bail!("unexpected version of ARK cert");
+    }
+
+    if ask.tbs_certificate.version != Version::V3 {
+        anyhow::bail!("unexpected version of ASK cert");
+    }
+
+    verify_cert_signature(ark, ask)?;
+    verify_cert_signature(ark, ark)?;
+
+    Ok(())
+}
+
+pub fn verify_cert_signature(signer: &Certificate, signee: &Certificate) -> anyhow::Result<()> {
+    if signee.signature_algorithm.oid != RSA_SSA_PSS_OID {
+        anyhow::bail!(
+            "unsupported signature algorithm: {:?}",
+            signee.signature_algorithm
+        );
+    }
+
+    let verifying_key = {
+        let pubkey_info = signer
+            .tbs_certificate
+            .subject_public_key_info
+            .owned_to_ref();
+        let pubkey = RsaPublicKey::try_from(pubkey_info)
+            .map_err(|_err| anyhow::anyhow!("could not parse RSA public key"))?;
+        rsa::pss::VerifyingKey::<Sha384>::new(pubkey)
+    };
+
+    let message = signee
+        .tbs_certificate
+        .to_der()
+        .map_err(|_err| anyhow::anyhow!("could not extract message to verify RSA signature"))?;
+    let signature = Signature::try_from(signee.signature.raw_bytes())
+        .map_err(|_err| anyhow::anyhow!("could not extract RSA signature"))?;
+
+    Ok(verifying_key
+        .verify(&message, &signature)
+        .map_err(|_err| anyhow::anyhow!("signature verification failed"))?)
+}
+
+fn to_array_64<T>(slice: &[T]) -> anyhow::Result<&[T; 64]> {
```

Comment:
I don't think this function is needed. Equality checks can be done on slices, so a slice over the `Vec` and a slice over the `[u8; 64]` can be compared.